### ISSUE TITLE
Very strange problem about training further with a pretrained model 

### ISSUE THREAD

**Issue by ErnstHowie** [dnouri/nolearn#128]
Hi everyone:

I was just following the great tutorial (https://github.com/mikesj-public/convolutional_autoencoder/blob/master/mnist_conv_autoencode.ipynb) about convolutional autoencoder using nolearn. 
I trained the defined NeuralNet (convolutional_autoencoder) for 100 epochs. I got a very good result: 0.003900 valid_loss. 
     97       [36m0.00393[0m       [32m0.00393[0m      1.00081  34.76s
     98       [36m0.00392[0m       [32m0.00392[0m      1.00004  33.57s
     99       [36m0.00391[0m       [32m0.00391[0m      1.00057  33.68s
    100       [36m0.00390[0m       [32m0.00390[0m      1.00028  34.55s
And the reconstruction images also look good, please see the following image:
![xpred_test](https://cloud.githubusercontent.com/assets/9126558/8843182/b277f9ac-30c2-11e5-9337-965b2fa6d274.png)

Then I want to keep on training further based on the pretrained model to see if a better results can be achieved. So I firstly loaded the pretrained model with the following code then trained more 10 epochs:

    with open(modelname_pretrained, 'rb') as f:
          pretrained_nn = pickle.load(f)
    if pretrained_nn is not None:
         ae.load_params_from(pretrained_nn)

I got the following output:
  epoch    train loss    valid loss    train/val  dur
-------  ------------  ------------  -----------  -------
      1       0.07188       0.06752      1.06456  217.32s
      2       0.06692       0.06721      0.99559  214.75s
      3       0.06677       0.06720      0.99357  211.39s
      4       0.06677       0.06719      0.99382  210.64s
      5       0.06678       0.06715      0.99439  211.10s
      6       0.06678       0.06707      0.99572  212.59s
      7       0.06679       0.06698      0.99723  212.03s
      8       0.06680       0.06694      0.99784  211.54s
      9       0.06683       0.06707      0.99644  210.28s
     10       0.06687       0.06697      0.99852  210.36s
![xpred_test](https://cloud.githubusercontent.com/assets/9126558/8843294/f42166f8-30c3-11e5-9582-7d8b84731a41.png)

Something is wrong:
1. the train/valid_loss should be close to 0.00390
2. the output reconstruction image shouldn't look like that (these are nothing)
I think these above results are very strange and unreasonable, but I don't know why. 
Anyone can help me? Thanks in advance!

**Comment by dnouri** [comment]
If used with a NeuralNet with a `verbose=1` parameter, `ae.load_params_from` should print some debugging info.  How does that look like?

**Comment by ErnstHowie** [comment]
Hello @dnouri ! The following is the output:

Loaded parameters to layer 'conv1' (shape 64x1x5x5).
Loaded parameters to layer 'conv1' (shape 64).
Loaded parameters to layer 'dense_encode' (shape 9216x128).
Loaded parameters to layer 'dense_encode' (shape 128).
Loaded parameters to layer 'dense_decode' (shape 128x16384).
Loaded parameters to layer 'dense_decode' (shape 16384).
Loaded parameters to layer 'deconv1' (shape 1x64x5x5).
Loaded parameters to layer 'deconv1' (shape 1).

The output looks like good! ae.load_params_from() tried to load the weights and biases, am I right? Or did I miss something?

**Comment by dnouri** [comment]
Hmm, that looks okay indeed.  Anything different about your training data?  Or maybe you're using a different conv layer implementation in your second training run?

**Comment by ErnstHowie** [comment]
The training data is same. It's the mnist dataset. And my conv layer implementations are also same.
I used the lasagne.layers.cuda_convnet  and layers.Conv2DLayer at the same time. Please see the following code.
Is that the problem?

        try:
		  from lasagne.layers.cuda_convnet import Conv2DCCLayer as Conv2DLayer
		  from lasagne.layers.cuda_convnet import MaxPool2DCCLayer as MaxPool2DLayer
        except ImportError:
		  Conv2DLayer = layers.Conv2DLayer
		  MaxPool2DLayer = layers.MaxPool2DLayer

         layers=[
			('input', layers.InputLayer),
			('conv1', Conv2DLayer),
			('pool1', MaxPool2DLayer),
			('flatten', ReshapeLayer),
			('dense_encode', layers.DenseLayer),
			('dense_decode', layers.DenseLayer),
			('unflatten', ReshapeLayer),
			('unpool1', Unpool2DLayer),
			('deconv1', layers.Conv2DLayer),   # for cuda_convnet, the number of filters must be a multiple of 16. So I have to use layers.Conv2DLayer
			('output', ReshapeLayer),
			],

**Comment by dnouri** [comment]
Did you switch between systems after the first round of training?  This does look like you might have had `Conv2DCCLayer` available for the first round of training but not the second.

**Comment by ErnstHowie** [comment]
Yes, you are very right! That is the problem. I run the first training on my lab's desktop systems and the second training on my laptop. Today I run this experiment on the same machine again. The results are reasonable. I never thought the difference between the lasagne.layers.cuda_convnet and layers.Conv2DLayer would make any problem. Thanks very much! @dnouri Nolearn is great. It is much easier to use than using Theano directly.

**Comment by RajkumarB** [comment]
Sorry for commenting on the closed thread. @ErnstHowie After loading pretrained model and all parameters using the following code,

with open(modelname_pretrained, 'rb') as f:
      pretrained_nn = pickle.load(f)
if pretrained_nn is not None:
     ae.load_params_from(pretrained_nn)

how to continue training for 10 more epochs?

I mean, how to pass these loaded values to the network defined as follows,

ae = NeuralNet(
    layers=[
        ('input', layers.InputLayer),
        ('conv', layers.Conv2DLayer),
        ('pool', layers.MaxPool2DLayer),
        ('flatten', ReshapeLayer), # output_dense
        ('encode_layer', layers.DenseLayer),
        ('hidden', layers.DenseLayer), # output_dense
        ('unflatten', ReshapeLayer),
        ('unpool', Unpool2DLayer),
        ('deconv', layers.Conv2DLayer),
        ('output_layer', ReshapeLayer),
        ],
    input_shape=(None, 1, 28, 28),
    conv_num_filters=conv_filters, conv_filter_size = (filter_sizes, filter_sizes), 
    conv_border_mode="valid",
    conv_nonlinearity=None,
    pool_ds=(2, 2),
    flatten_shape=(([0], -1)), # not sure if necessary?
    encode_layer_num_units = encode_size,
    hidden_num_units= deconv_filters * (28 + filter_sizes - 1) ** 2 / 4,
    unflatten_shape=(([0], deconv_filters, (28 + filter_sizes - 1) / 2, (28 + filter_sizes - 1) / 2 )),
    unpool_ds=(2, 2),
    deconv_num_filters=1, deconv_filter_size = (filter_sizes, filter_sizes), 
    deconv_border_mode="valid",
    deconv_nonlinearity=None,
    output_layer_shape = (([0], -1)),
    update_learning_rate = 0.01,
    update_momentum = 0.975,
    batch_iterator_train=FlipBatchIterator(batch_size=128),
    regression=True,
    max_epochs= epochs,
    verbose=1,
    )

**Comment by ErnstHowie** [comment]
@RajkumarB You need to define a NeuralNet with the above values but has the same structure as the pretrained model. Then using " ae.load_params_from(pretrained_nn)", the pretrained params(weights and biases) will be loaded and used to initialize the defined NeuralNet.

**Comment by RajkumarB** [comment]
Thank you very much @ErnstHowie
